Modules/arraymodule.c

```diff
@@ -728,19 +1006,26 @@ array_dealloc(PyObject *op)
     PyObject_GC_UnTrack(op);
 
     arrayobject *self = arrayobject_CAST(op);
+    if (self->ob_exports > 0) {
+        PyErr_SetString(PyExc_SystemError,
+                        "deallocated array object has exported buffers");
+        PyErr_WriteUnraisable(NULL);
+    }
     if (self->weakreflist != NULL) {
         PyObject_ClearWeakRefs(op);
     }
-    if (self->ob_item != NULL) {
-        PyMem_Free(self->ob_item);
+    if (self->data != NULL) {
+        arraydata_free(self->data, false);
```

Comment:
The `if` isn't necessary. Just `arraydata_free(self->data, false);` should be fine.

Reply:
Will remove, but just to double check I added it because I thought the NULL can get here:

https://github.com/tom-pytel/cpython/blob/c5abded09995f208b21ebaf012185ca5acb0180b/Objects/obmalloc.c#L1237

Reply:
Hmm, you might be right. It's okay to leave it in. We should probably make `_PyMem_FreeDelayed` handle `NULL` pointers for consistency with C's `free()` and `PyMem_Free()`, but we don't have to do that now.

Reply:
Added PR for this: #131720, if you merge that first I will remove the check here.